### PR TITLE
Electra: upgrade

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**

- `NewGenesisBlockForState` returns Electra state
- `UpgradeToElectra` upgrades Deneb state to Electra
- `CanUpgradeToElectra` returns true if it's Electra epoch
- `ProcessSlots` and `ReplayProcessSlots` can upgrade to Electra
- Add spec test for upgrade

**Which issues(s) does this PR fix?**

part of PSM-371

**Other notes for review**

dependency

- [x] https://github.com/prysmaticlabs/prysm/pull/13919

### CODE REVIEW

File: beacon-chain/core/electra/upgrade.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	var preActivation []primitives.ValidatorIndex
+
+	for index, validator := range s.Validators {
+		if validator.ActivationEpoch == params.BeaconConfig().FarFutureEpoch {
+			preActivation = append(preActivation, primitives.ValidatorIndex(index))
+		}
+	}
+
+	// Sorting preActivation based on a custom criteria
+	sort.Slice(preActivation, func(i, j int) bool {
```
Comment (james-prysm) [inline beacon-chain/core/electra/upgrade.go:185]: This part of the code needs more unit testing

File: beacon-chain/core/electra/upgrade_test.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+	startIndex, err := mSt.DepositReceiptsStartIndex()
+	require.NoError(t, err)
+	require.Equal(t, params.BeaconConfig().UnsetDepositReceiptsStartIndex, startIndex)
+
+	balance, err := mSt.DepositBalanceToConsume()
+	require.NoError(t, err)
+	require.Equal(t, math.Gwei(0), balance)
+
+	//TODO: need getters for ExitBalanceToConsume and EarliestExitEpoch to test
```
Comment (james-prysm) [inline beacon-chain/core/electra/upgrade_test.go:153]: TODO here

File: config/params/mainnet_config.go [hunk 3]
```diff
@@ -290,6 +290,7 @@ var mainnetBeaconConfig = &BeaconChainConfig{
 	MaxPendingPartialsPerWithdrawalsSweep: 8,
 	FullExitRequestAmount:                 0,
 	MaxWithdrawalRequestsPerPayload:       16,
+	UnsetDepositReceiptsStartIndex:        math.MaxUint64,
```
Comment (james-prysm) [inline config/params/mainnet_config.go:293]: double check if constant should go here?

File: testing/spectest/shared/electra/fork/upgrade_to_electra.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+			postStateSSZ, err := snappy.Decode(nil /* dst */, postStateFile)
+			require.NoError(t, err, "Failed to decompress")
+			postStateFromFile := &ethpb.BeaconStateElectra{}
+			if err := postStateFromFile.UnmarshalSSZ(postStateSSZ); err != nil {
+				t.Fatalf("Failed to unmarshal: %v", err)
+			}
+
+			if !proto.Equal(postStateFromFile, postStateFromFunction) {
+				t.Log(postStateFromFile.LatestExecutionPayloadHeader)
+				t.Log(postStateFromFunction.LatestExecutionPayloadHeader)
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply testing/spectest/shared/electra/fork/upgrade_to_electra.go:55]: Yeah this doesn't make sense. Please copy the upgrade_to_electra.go files from #13903 where we can log the proto diffs

```go
t.Log(cmp.Diff(postStateFromFile, postStateFromFunction, protocmp.Transform())
```

[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/spectest/shared/electra/fork/upgrade_to_electra.go:55]: will double check this

File: beacon-chain/core/electra/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	s := &ethpb.BeaconStateElectra{
+		GenesisTime:           state.GenesisTime(),
+		GenesisValidatorsRoot: state.GenesisValidatorsRoot(),
+		Slot:                  state.Slot(),
+		Fork: &ethpb.Fork{
+			PreviousVersion: state.Fork().CurrentVersion,
+			CurrentVersion:  params.BeaconConfig().ElectraForkVersion,
+			Epoch:           epoch,
```
Comment (prestonvanloon) [inline beacon-chain/core/electra/upgrade.go:114]:
```suggestion
	currentSyncCommittee, err := state.CurrentSyncCommittee()
	if err != nil {
		return nil, err
	}
	nextSyncCommittee, err := state.NextSyncCommittee()
	if err != nil {
		return nil, err
	}
	prevEpochParticipation, err := state.PreviousEpochParticipation()
	if err != nil {
		return nil, err
	}
	currentEpochParticipation, err := state.CurrentEpochParticipation()
	if err != nil {
		return nil, err
	}
	inactivityScores, err := state.InactivityScores()
	if err != nil {
		return nil, err
	}
	payloadHeader, err := state.LatestExecutionPayloadHeader()
	if err != nil {
		return nil, err
	}
	txRoot, err := payloadHeader.TransactionsRoot()
	if err != nil {
		return nil, err
	}
	wdRoot, err := payloadHeader.WithdrawalsRoot()
	if err != nil {
		return nil, err
	}
	wi, err := state.NextWithdrawalIndex()
	if err != nil {
		return nil, err
	}
	vi, err := state.NextWithdrawalValidatorIndex()
	if err != nil {
		return nil, err
	}
	summaries, err := state.HistoricalSummaries()
	if err != nil {
		return nil, err
	}
	historicalRoots, err := state.HistoricalRoots()
	if err != nil {
		return nil, err
	}
	excessBlobGas, err := payloadHeader.ExcessBlobGas()
	if err != nil {
		return nil, err
	}
	blobGasUsed, err := payloadHeader.BlobGasUsed()
	if err != nil {
		return nil, err
	}

	// RTFM: https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/fork.md
	// Find the earliest exit epoch
	exitEpochs := make([]primitives.Epoch, 0)
	for _, v := range state.Validators() {
		if v.ExitEpoch != params.BeaconConfig().FarFutureEpoch {
			exitEpochs = append(exitEpochs, v.ExitEpoch)
		}
	}
	if len(exitEpochs) == 0 {
		exitEpochs = append(exitEpochs, time.CurrentEpoch(state))
	}
	var earliestExitEpoch primitives.Epoch
	for _, e := range exitEpochs {
		if e > earliestExitEpoch {
			earliestExitEpoch = e
		}
	}
	earliestExitEpoch++ // Increment to find the earliest possible exit epoch

	// note: should be the same in prestate and post state.
	// we are deviating from the specs a bit as it calls for using the post state
	tab, err := helpers.TotalActiveBalance(state)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get total active balance")
	}

	s := &ethpb.BeaconStateElectra{
		GenesisTime:           state.GenesisTime(),
		GenesisValidatorsRoot: state.GenesisValidatorsRoot(),
		Slot:                  state.Slot(),
		Fork: &ethpb.Fork{
			PreviousVersion: state.Fork().CurrentVersion,
			CurrentVersion:  params.BeaconConfig().ElectraForkVersion,
			Epoch:           time.CurrentEpoch(state),
```

No need for this epoch variable

File: beacon-chain/core/electra/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	excessBlobGas, err := payloadHeader.ExcessBlobGas()
+	if err != nil {
+		return nil, err
+	}
+	blobGasUsed, err := payloadHeader.BlobGasUsed()
+	if err != nil {
+		return nil, err
+	}
+
+	// RTFM: https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/fork.md
```
Comment (prestonvanloon) [inline beacon-chain/core/electra/upgrade.go:81]:
```suggestion
```
Looks like this came from #13903 and this comment was a note to re-read the spec to ensure it is compliant. aka this was a note to myself :rofl:

File: beacon-chain/core/electra/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	// Find the earliest exit epoch
+	exitEpochs := make([]primitives.Epoch, 0)
+	for _, v := range state.Validators() {
+		if v.ExitEpoch != params.BeaconConfig().FarFutureEpoch {
+			exitEpochs = append(exitEpochs, v.ExitEpoch)
+		}
+	}
```
Comment (prestonvanloon) [inline beacon-chain/core/electra/upgrade.go:88]: Prefer `ReadFromEveryValidator(f func(idx int, val ReadOnlyValidator) error) error`, it doesn't make a copy of the validator set as you have here.

File: beacon-chain/core/electra/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+	// add validators that are not yet active to pending balance deposits
+
+	// Creating a slice to store indices of validators whose activation epoch is set to FAR_FUTURE_EPOCH
+	var preActivation []primitives.ValidatorIndex
+
+	for index, validator := range s.Validators {
+		if validator.ActivationEpoch == params.BeaconConfig().FarFutureEpoch {
+			preActivation = append(preActivation, primitives.ValidatorIndex(index))
+		}
+	}
```
Comment (prestonvanloon) [inline beacon-chain/core/electra/upgrade.go:182]: Same comment here about `ReadFromEveryValidator`. Can you collect both pieces of information in the same `ReadFromEveryValidator` call?

File: beacon-chain/core/electra/validator.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+//	def queue_entire_balance_and_reset_validator(state: BeaconState, index: ValidatorIndex) -> None:
+//	    balance = state.balances[index]
+//	    validator = state.validators[index]
+//		state.balances[index] = 0
+//	    validator.effective_balance = 0
+//	    validator.activation_eligibility_epoch = FAR_FUTURE_EPOCH
+//	    state.pending_balance_deposits.append(
+//	        PendingBalanceDeposit(index=index, amount=balance)
+//	    )
+func QueueEntireBalanceAndResetValidator(s state.BeaconState, idx primitives.ValidatorIndex) error {
```
Comment (prestonvanloon) [inline beacon-chain/core/electra/validator.go:22]: Missing tests?

File: beacon-chain/state/stategen/replay.go [hunk 10]
```diff
@@ -201,20 +202,19 @@ func ReplayProcessSlots(ctx context.Context, state state.BeaconState, slot primi
...
+			} else if state.Version() >= version.Altair {
 				state, err = altair.ProcessEpoch(ctx, state)
 				if err != nil {
 					tracing.AnnotateError(span, err)
 					return nil, errors.Wrap(err, "could not process epoch")
 				}
-			default:
+			} else {
 				return nil, fmt.Errorf("unsupported beacon state version: %s", version.String(state.Version()))
 			}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/state/stategen/replay.go:219]: can remove

File: beacon-chain/core/electra/upgrade.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+	// Ensure early adopters of compounding credentials go through the activation churn
+	for index, validator := range s.Validators {
+		if helpers.HasCompoundingWithdrawalCredential(validator) {
+			if err := QueueEntireBalanceAndResetValidator(post, primitives.ValidatorIndex(index)); err != nil {
+				return nil, errors.Wrap(err, "failed to queue entire balance and reset validator")
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go:211]: can try to do that

[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go:211]: I think this needs to be a separate loop, since queue Entire balance needs to run before and I need the initialized proto unsafe electra

File: beacon-chain/core/electra/upgrade.go (hunk 8 above)
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go:181]: moved them together

File: beacon-chain/core/electra/validator.go (hunk 9 above)
Comment (james-prysm) [reply beacon-chain/core/electra/validator.go:22]: took from maxeb, forgot to add test

File: beacon-chain/core/electra/upgrade.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go]: nvm I merged them, however I am now wondering if my added function should be there or if the helpers should be changed

File: beacon-chain/core/electra/upgrade.go [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+	// get all the validators with compound withdrawal indicies for eip7521
+	compoundWithdrawalIndices := make([]primitives.ValidatorIndex, 0)
+	if err = beaconState.ReadFromEveryValidator(func(index int, val state.ReadOnlyValidator) error {
+		if val.ExitEpoch() != params.BeaconConfig().FarFutureEpoch {
+			exitEpochs = append(exitEpochs, val.ExitEpoch())
+		}
+		if val.ActivationEpoch() == params.BeaconConfig().FarFutureEpoch {
+			preActivationIndices = append(preActivationIndices, primitives.ValidatorIndex(index))
+		}
+		if helpers.HasCompoundingWithdrawalCredential(val.Validator()) {
```
Comment (james-prysm) [inline beacon-chain/core/electra/upgrade.go:206]: should this function `val.Validator()` be added? or maybe helpers.HasCompounderWithdrawalCredentials should be changed to accept the readonly interface?

File: beacon-chain/core/electra/upgrade.go (hunk 5 above)
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go]: updated

File: beacon-chain/core/electra/upgrade.go (hunk 12 above)
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go]: changed to use interface

File: beacon-chain/core/electra/upgrade.go [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+
+	for _, index := range preActivationIndices {
+		if err := QueueEntireBalanceAndResetValidator(post, index); err != nil {
+			return nil, errors.Wrap(err, "failed to queue entire balance and reset validator")
+		}
+	}
+
+	// Ensure early adopters of compounding credentials go through the activation churn
+	for _, index := range compoundWithdrawalIndices {
+		if err := QueueEntireBalanceAndResetValidator(post, primitives.ValidatorIndex(index)); err != nil {
```
Comment (james-prysm) [inline beacon-chain/core/electra/upgrade.go]: using the wrong function here... will fix

File: testing/util/electra_state.go [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+		// Eth1 data.
+		Eth1Data:         eth1Data,
+		Eth1DataVotes:    []*ethpb.Eth1Data{},
+		Eth1DepositIndex: preState.Eth1DepositIndex(),
+
+		//Electra Data
+		DepositReceiptsStartIndex:     params.BeaconConfig().UnsetDepositReceiptsStartIndex,
+		ExitBalanceToConsume:          helpers.ActivationExitChurnLimit(math.Gwei(tab)),
+		EarliestConsolidationEpoch:    helpers.ActivationExitEpoch(slots.ToEpoch(preState.Slot())),
+		ConsolidationBalanceToConsume: helpers.ConsolidationChurnLimit(math.Gwei(tab)),
```
Comment (james-prysm) [inline testing/util/electra_state.go:193]: wondering how important to set value of earliest exit here

File: beacon-chain/core/electra/upgrade.go [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+	if err != nil {
+		return nil, err
+	}
+	blobGasUsed, err := payloadHeader.BlobGasUsed()
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the earliest exit epoch
+	exitEpochs := make([]primitives.Epoch, 0)
```
Comment (terencechain) [inline beacon-chain/core/electra/upgrade.go:193]: You dont need a list here. You can save on memory. Just track a single variable of `primitives.Epoch`, first set it to `time.CurrentEpoch(beaconState)`, then override it if there's a higher one. Finally, increment it by one

File: beacon-chain/core/electra/upgrade.go [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+	}
+
+	// Sorting preActivationIndices based on a custom criteria
+	sort.Slice(preActivationIndices, func(i, j int) bool {
+		// Comparing based on ActivationEligibilityEpoch and then by index if the epochs are the same
+		if s.Validators[preActivationIndices[i]].ActivationEligibilityEpoch == s.Validators[preActivationIndices[j]].ActivationEligibilityEpoch {
+			return preActivationIndices[i] < preActivationIndices[j]
+		}
+		return s.Validators[preActivationIndices[i]].ActivationEligibilityEpoch < s.Validators[preActivationIndices[j]].ActivationEligibilityEpoch
+	})
```
Comment (terencechain) [inline beacon-chain/core/electra/upgrade.go:305]: Why do we need to sort here? I know spec says it but why?  We should avoid wasteful compute if we can

**Review by terencechain — COMMENTED** [review]
I only reviewed `UpgradeToElectra`. Happy to review the helpers tomorrow, just let me know

File: beacon-chain/core/electra/upgrade.go [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+	}
+	blobGasUsed, err := payloadHeader.BlobGasUsed()
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the earliest exit epoch
+	exitEpochs := make([]primitives.Epoch, 0)
+	// [New in Electra:EIP7251]
+	// add validators that are not yet active to pending balance deposits
```
Comment (rkapka) [inline beacon-chain/core/electra/upgrade.go:195]: This comment feels out of place

File: beacon-chain/state/interfaces.go [hunk 18]
```diff
@@ -115,7 +116,7 @@ type ReadOnlyValidator interface {
 	WithdrawableEpoch() primitives.Epoch
 	ExitEpoch() primitives.Epoch
 	PublicKey() [fieldparams.BLSPubkeyLength]byte
-	WithdrawalCredentials() []byte
+	GetWithdrawalCredentials() []byte
```
Comment (rkapka) [inline beacon-chain/state/interfaces.go:119]: Why the change?

File: beacon-chain/state/interfaces.go [hunk 19]
```diff
@@ -179,6 +180,12 @@ type ReadOnlyEth1Data interface {
 	Eth1DepositIndex() uint64
 }
 
+// ReadOnlyExits defines a struct which only has read access to Exit related methods.
+type ReadOnlyExits interface {
+	ExitBalanceToConsume() (math.Gwei, error)
+	EarliestExitEpoch() (primitives.Epoch, error)
+}
```
Comment (rkapka) [inline beacon-chain/state/interfaces.go:187]: This stuff is never used.

File: beacon-chain/state/stategen/replay.go [hunk 20]
```diff
@@ -201,21 +203,18 @@ func ReplayProcessSlots(ctx context.Context, state state.BeaconState, slot primi
...
-			switch state.Version() {
-			case version.Phase0:
+			if state.Version() == version.Phase0 {
 				state, err = transition.ProcessEpochPrecompute(ctx, state)
 				if err != nil {
 					tracing.AnnotateError(span, err)
 					return nil, errors.Wrap(err, "could not process epoch with optimizations")
 				}
-			case version.Altair, version.Bellatrix, version.Capella, version.Deneb:
+			} else {
```
Comment (rkapka) [inline beacon-chain/state/stategen/replay.go:212]: Nice :+1:

File: testing/util/electra_state.go [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+	fieldparams "github.com/prysmaticlabs/prysm/v5/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v5/config/params"
+	"github.com/prysmaticlabs/prysm/v5/crypto/bls"
+	"github.com/prysmaticlabs/prysm/v5/math"
+	enginev1 "github.com/prysmaticlabs/prysm/v5/proto/engine/v1"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/time/slots"
+)
+
+// DeterministicGenesisStateElectra returns a genesis state in Deneb format made using the deterministic deposits.
```
Comment (rkapka) [inline testing/util/electra_state.go:21]:
```suggestion
// DeterministicGenesisStateElectra returns a genesis state in Electra format made using the deterministic deposits.
```

File: beacon-chain/state/interfaces.go (hunk 18 above)
Comment (james-prysm) [reply beacon-chain/state/interfaces.go:119]: It needs to follow the interface

File: testing/util/electra_state.go [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+		Eth1Data:         &ethpb.Eth1Data{},
+		Eth1DataVotes:    []*ethpb.Eth1Data{},
+		Eth1DepositIndex: 0,
+
+		LatestExecutionPayloadHeader: &enginev1.ExecutionPayloadHeaderElectra{},
+	}
+	return state_native.InitializeFromProtoElectra(st)
+}
+
+func buildGenesisBeaconStateElectra(genesisTime uint64, preState state.BeaconState, eth1Data *ethpb.Eth1Data) (state.BeaconState, error) {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply testing/util/electra_state.go:90]: will try that

File: consensus-types/interfaces/validator.go [hunk 23]
```diff
@@ -0,0 +1,5 @@
+package interfaces
+
+type WithWithdrawalCredentials interface {
+	GetWithdrawalCredentials() []byte
+}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply consensus-types/interfaces/validator.go:5]: it's used in the helper function so that readOnly validator and protovalidator both can be passed in.

File: beacon-chain/state/interfaces.go (hunk 19 above)
Comment (james-prysm) [reply beacon-chain/state/interfaces.go:187]: used in tests

Comment (rkapka) [reply beacon-chain/state/interfaces.go:187]: in tests you can just cast to the appropriate type and read the field directly

File: beacon-chain/state/interfaces.go (hunk 18 above)
Comment (rkapka) [reply beacon-chain/state/interfaces.go:119]: the `WithWithdrawalCredentials` interface? I think that interface's function can also be called `WithdrawalCredentials` and it shouldn't cause an issue

Comment (james-prysm) [reply beacon-chain/state/interfaces.go:119]: the protobuf doesn't have WithdrawalCredentials as a function, only GetWithdrawalCredentials

File: beacon-chain/core/electra/upgrade.go (hunk 16 above)
Comment (james-prysm) [reply beacon-chain/core/electra/upgrade.go:305]: validators might have different ActivationEligibilityEpoch and should be sorted based on that value, there is also a tie breaker situation. i don't have a good solution to replace this right now.

File: beacon-chain/rpc/eth/config/handlers_test.go [hunk 24]
```diff
@@ -40,6 +40,7 @@ func TestGetDepositContract(t *testing.T) {
 	assert.Equal(t, "0x4242424242424242424242424242424242424242", response.Data.Address)
 }
 
+// skipcq: GO-R1005
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/config/handlers_test.go:43]: What is this?

**Review by rkapka — APPROVED** [review]

**Review by prestonvanloon — APPROVED** [review]
